### PR TITLE
Typed `nixpkgs.config` married to NixOS

### PR DESCRIPTION
**Pull request by oxij**
# What? Why?

This is the next completely ready piece of #56227.

The four commits from master proceeding this branch, namely 570aed4b46916551da969c4daeef0322461ac2fe 4a647dd225571e91d81c1306c69e3473fdd9a8c5 5aa813dcc942d05485b89a89e734d231f31eaa9f 83ae1ffed43f5dc54bcfd58655974293d215df6e, are the previous completely ready piece cherry-picked onto master from #56227 by @danbst.

The previous piece introduced option checking to `pkgs`, this piece marries it to `nixpkgs.config` option of NixOS.

To make this marriage possible I had to type `packageOverrides` in `pkgs` (to remove ad-hoc processing of `nixpkgs.config` in the corresponding NixOS module). To type `packageOverrides` I had to revive `types.functionTo`.

# `git log`

- check-meta.nix: move options to `config.nix`

  This also makes "Unfree" and "Insecure" options homogenous.

- lib: resurrect `types.functionTo`

  This reverts commit 4ff1ab5a56f1280d2de319ad4eb4b2796e07ed35.

  This type was removed because it usually implies a bad design. However, since that
  commit NixOS modules that can't live without it implemented several ugly ad-hoc
  replacements for it while other modules that could make a good use of it
  simply suffer. See the following commits.

- pkgs/top-level/config.nix: add various `packageOverrides`

  Hm, Haskell's one is actually an overlay.

- lib/modules.nix: rename _file -> file in unifyModuleSyntax

  ... and change all of its uses.

  Strangely enough, before this `_file` was in the public interface of the module
  system while `file` was used internally by `lib.evalModules` code
  after `lib.unifyModuleSyntax` was applied.

  `_file` attribute was introduced in 800f9c203728ee84d50181af6fa34ac9e5c04e73 but
  I see no reason behind having "_" in the name there, all other attrs
  processed by `lib.unifyModuleSyntax` keep their names between its input and output.

  On the other hand, renaming `_file` into `file` makes `lib.unifyModuleSyntax`
  accept its own output which is useful in cases when you want to modify
  modules before shoving them into `lib.evalModules` but you don't want to duplicate
  `lib.unifyModuleSyntax`.

- lib/modules.nix: make unifyModuleSyntax idempotent

  The previous patch made it accept its own output, this also makes it
  idempotent.

- lib/modules.nix: simplify pushDownProperties

- lib/types.nix: add opaque type

- pkgs/top-level/impure.nix: move defaults into the preceding `let` block

- pkgs/top-level: introduce `configs` argument and marry `nixpkgs.config` of NixOS to it

  This types `nixpkgs.config` of NixOS with `types.opaque` type and applies the
  resulting evaluated value to the newly introduced `configs` argument of `pkgs`.

  The new `configs` argument is like the old `config` argument (still available for
  convenience) but a list.

  This design was chosen because we can't just make a submodule out of
  `pkgs/top-level/config.nix` in NixOS because

  - firstly, we would have to duplicate all the magic `pkgs/top-level/default.nix`
    does to `config` in NixOS,

  - secondly, in principle, we might not have all the arguments needed by
    `pkgs/top-level/config.nix` in NixOS because some of them can be
    computed internally by `pkgs` (this is not a problem now
    but it will be a problem in use-flags patchset that comes after), thus all
    of those computations will need to be duplicated too,

  - thirdly, doing it this way keeps NixOS and pkgs separated, so that, in
    principle, one could replace NixOS with an alternative thing without duplicating
    all of the above once again and/or one could use separate module and type
    systems between NixOS and pkgs (`types.opaque` is trivial to
    implement in any type system and pkgs can use "types-simple" of #37252 today,
    if we were to wish it so).

  Note that since this design removes the need to do any preprocessing of
  `nixpkgs.config` in NixOS all the ad-hoc merging code was removed. Yay!

# `nix-instantiate` environment

- Host OS: Linux 4.9, SLNOS 19.09
- Nix: nix-env (Nix) 2.2
- Multi-user: yes
- Sandbox: yes
- NIXPKGS_CONFIG:

```nix
{
  checkMeta = true;
  doCheckByDefault = true;
}
```

# `nix-env -qaP` diffs

- On x86_64-linux:
  - Updated (1):
    - tests.nixos-functions.nixos-test
- On aarch64-linux: ditto
- On x86_64-darwin: noop

/cc @dtzWill @Ericson2314 @matthewbauer @7c6f434c @Infinisil @edolstra from the previous PR

### CODE REVIEW

**Comment by oxij** [comment]
What to pay attention to while reviewing:

- types in `config.nix`, especially the ones for `*packageOverrides`, I think they are correct, but I might have missed something,
- do commit messages make sense?

**Comment by oxij** [comment]
Btw, changes in metrics of `nix-env -f . -qaP --drv-path --out-path` between the parent of 570aed4b46916551da969c4daeef0322461ac2fe and this HEAD:

Before:

```
time 47.205 s
maxresident 4264000 KiB
allocations 6007093832 B
values 80800261
```

After:

```
time 47.047 s
maxresident 4241936 KiB
allocations 5956525992 B
values 79612951
```

I wouldn't believe the time metric, it's too random, but the rest are most certainly valid.

So, despite adding type checking to `nixpkgs.config` I _improved_ all the things :]

**Comment by oxij** [comment]
Hm, also 1d1ebf5a58a1ad1d87a8fa067ee0c56f8b9a87e1 changes the semantics of `haskellPackageOverrides` a bit as `config.haskellPackageOverrides` is never undefined now. @peti, should I change that? If so then how? The interaction between `config.haskellPackageOverrides` and `haskell.packageOverrides` seems nontrivial.

**Comment by danbst** [comment]
### haskellPackageOverrides
As for `haskellPackageOverrides`, it is better to comment it. Maybe it is deprecated by overlays? It will also remove the need for @peti to review this PR.

### functionTo
Do we need generic "function to"? I'm biased here, as I like my "dependently typed selector" variant (https://github.com/NixOS/nixpkgs/pull/44601#issuecomment-470908855). `listSelector`, `attrsSelector` and `overlay` are the only required function types. ... I'm not sure overlays are worth to encode in types.

### _file -> file
I've used `_file` for my custom modules already. There has to be backwards compat. But I also wonder why was it named like that. Maybe, to discourage usage? In that case I have a valid usecase:
```
let
  module1 = {
     key = "module1"; _file = "this_file.nix";
     config = { error.option = true; };
  };
  module2 = {
     key = "module2"; _file = "this_file.nix";
  };
in {
  imports = [ module1 module2 ];
}
```
```
# without _file
nix-repl> evalModules { modules = [ (import ../simple-2-modules.nix) ]; }
{ config = «error: The option `error' defined in `<unknown-file>' does not exist.»; options = { ... }; }

# with _file
nix-repl> evalModules { modules = [ (import ../simple-2-modules.nix) ]; }
{ config = «error: The option `error' defined in `this_file.nix' does not exist.»; options = { ... }; }
```
At least I know file name.

### make unifyModuleSyntax idempotent 
> The previous patch made it accept its own output, this also makes it
idempotent.

I don't understand the comment. Which previous patch? As for changes, looks like you've removed `mkMerge` in cases when there is no `meta`, which is nice :+1: 

### pushDownProperties

:+1: I've also discovered that inside module system sometimes code is duplicated to not introduce extra variables, which take memory. Can you verify that it isn't the case here?

### opaque type
This looks like introspection thing. It would generate questions - "what is this?". I guess a "module system manual" is needed now. I'm also not a фанат of introducing concepts without usages - maybe combine into one commit with next one?

### move defaults
I'm okayish here, just a note that it makes you an author in `git blame`. @7c6f434c do we care about `git blame` when moving large pieces of code?

### the last commit
looks very strange, have yet to review it.

**Comment by oxij** [comment]
> ### haskellPackageOverrides

Yep, I guess, if @peti won't answer in the next couple of days I'll just comment it out.

> ### functionTo

I saw that, I'm :+1: on your proposal there but I think `packageOverrides` has exactly the type I gave it here and that can't be expressed as `*Selector`, AFAICS.

> ### _file -> file
>
> ...
> At least I know file name.

By pasting your example into a random config I get

> Module `configuration.nix:anon-1' has an unsupported attribute `_file'. This is caused by assignments to the top-level attributes `config' or `options'.

which is the expected correct behavior. You don't have a filename in your example because you `evalModules` on a nix expression, try to supply a path instead, e.g.

nix-repl> evalModules { modules = [ ../simple-2-modules.nix ]; }

> ### make unifyModuleSyntax idempotent 
>
> I don't understand the comment. Which previous patch?

_file -> file (duh)

> ### pushDownProperties

I don't believe it can make a difference in this case because the function is small and the value is immediately GC'd, but I'll check.

> ### opaque type

> This looks like introspection thing. It would generate questions - "what is this?". I guess a "module system manual" is needed now.

I added some explanations both to the comment above it and to the nixos manual, though.

> I'm also not a фанат of introducing concepts without usages - maybe combine into one commit with next one?

The commit that uses it is too big, IMHO.

> ### move defaults

What, you don't have

```
[alias]
    praise = blame -w -M -C
```

in your `~/.gitconfig`, such blasphemy!

> ### the last commit
> looks very strange, have yet to review it.

Yep, it touches a bunch of stuff, unfortunately I didn't come up with a way to break it into pieces.

**Comment by danbst** [comment]
>  Module `configuration.nix:anon-1' has an unsupported attribute `_file'

that is because it is a top-level module parameter. You can't place it into random config, you have to massage module first.

> You don't have a filename in your example because you `evalModules` on a nix expression, try to supply a path instead

You're right. But in my case I did exactly `import`. Another example:
```
# simple-module-generator.nix
{
    module1 = {
        error = true;
    };
    module2 = {
    };
}
```
```
# simple-2-modules.nix
with import ./simple-module-generator.nix;
{
  imports = [ module1 module2 ];
}
```
```
nix-repl> evalModules { modules = [ ../simple-2-modules.nix ]; }
{ config = «error: The option `error' defined in `/home/danbst/dev/simple-2-modules.nix' does not exist.»; options = { ... }; }
```
Error is not in `simple-2-modules.nix` file.

**Comment by oxij** [comment]
> ```
> # simple-2-modules.nix
> with import ./simple-module-generator.nix;
> {
>   imports = [ module1 module2 ];
> }
> ```
> ```
> nix-repl> evalModules { modules = [ ../simple-2-modules.nix ]; }
> { config = «error: The option `error' defined in `/home/danbst/dev/simple-2-modules.nix' does not exist.»; options = { ... }; }
> ```
> Error is not in `simple-2-modules.nix` file.

Well, sure, but it works the same on master and I don't see how this can be fixed using the current NixOS modules infra, if you don't specify `file` nor import the module via a path so that `evalModules` would set `file` for you it will be a "best guess", so to say.

**Comment by peti** [comment]
@oxij,

> 1d1ebf5a58a1ad1d87a8fa067ee0c56f8b9a87e1 changes the semantics of `haskellPackageOverrides` a bit as `config.haskellPackageOverrides` is never undefined now. @peti, should I change that? If so then how? The interaction between `config.haskellPackageOverrides` and `haskell.packageOverrides` seems nontrivial.

I removed `haskellPackageOverrides` entirely in ee1f34e9f1d89735d3ed78f8429d8435b807a247. Does that help?

**Comment by oxij** [comment]
@peti Yes, that's a possible solution :)

But in the meantime I was thinking that maybe it would make sense to make `haskellPackageOverrides` into the default for `haskell.packageOverrides`. E.g. suppose you want to override several `ghc` versions at the same time.

**Comment by danbst** [comment]
> Well, sure, but it works the same on master and I don't see how this can be fixed using the current NixOS modules infra, if you don't specify `file` nor import the module via a path so that `evalModules` would set `file` for you it will be a "best guess", so to say.

```
{
    module1 = {
        _file = "module at ${__curPos.file}:${toString __curPos.line}";
        config = {
            error = true;
        };
    };

    module2 = {
        _file = "module at ${__curPos.file}:${toString __curPos.line}";
    };
}
```
```
nix-repl> evalModules { modules = [ (import ../simple-2-modules.nix) ]; }
{ config = «error: The option `error' defined in `module at /home/danbst/dev/simple-module-generator.nix:3' does not exist.»; options = { ... }; }
```

But I see now that `key` and `_file` are not public module interface, nowhere in manual is said that you can use anonymous modules and should set key/_file explicitly. I'd vote for at least release notes then.

**Comment by peti** [comment]
> But in the meantime I was thinking that maybe it would make sense to make `haskellPackageOverrides` into the default for `haskell.packageOverrides`. E.g. suppose you want to override several `ghc` versions at the same time.

I am sorry, but I don't understand what you mean.

**Comment by oxij** [comment]
@danbst Mention in release notes sounds good.

@peti Ah, `haskell.packageOverrides` already overrides for all versions, nevermind me then.

**Comment by oxij** [comment]
@peti

> haskellPackageOverrides

See the new patch. I made your ee1f34e9f1d89735d3ed78f8429d8435b807a247 not as violent with the HEAD of this branch. Does that look ok?

@danbst

> release notes

Done.

> lib/modules.nix: simplify pushDownProperties

Removed. Indeed that produced an overall negative effect of the metrics. I suspect nix needs tail call optimization to make that change useful.

**Comment by peti** [comment]
> haskellPackageOverrides
> See the new patch. I made your [ee1f34e](https://github.com/NixOS/nixpkgs/commit/ee1f34e9f1d89735d3ed78f8429d8435b807a247) not as violent with the HEAD of this branch. Does that look ok?

No, please don't re-introduce that option. I'm glad it is gone.

I also notice there's an `rPackageOverrides` now which I don't want either (or did I misread the patch?).

**Comment by danbst** [comment]
@oxij Your thoughts about overlays in config are very interesting!

**Comment by danbst** [comment]
So, I'm reading the finalizing commit... I have two questions:
1. Maybe it would be cleaner to support `imports` directive in config? Think of:
```
{
  imports = [ ./my-custom-config-modules.nix ];

  allowBroken = true;  
}
```
Then a) it would be possible to add use-flags thing without amending NixOS code, only `extraScope` needed; b) it would remove the need to pass list of configs, just pass `config = { imports = cfg.config; }`

As for implementation, I think it would require some massaging of config, but should be doable.

2. It also not very cool that `pkgs/top-level/default.nix` includes hacky "check only known options". Yes, I agree this is sole usage, but it may be harder to sell to others. The `unknowns` hack is because module system doesn't support partial checks, right?

**Comment by oxij** [comment]
> 1. Maybe it would be cleaner to support `imports` directive in config?

In fact, this is supported already, config can be a normal NixOS module:

```nix
{ pkgs, ... }:
{
  imports = [
    ./nothing-interesting-here.nix
  ];

  config = {
    checkMeta = true;
    doCheckByDefault = false;
  };
}
```

> Then a) it would be possible to add use-flags thing without amending NixOS code, only `extraScope` needed;

Yes, though I think making a "flake" out of use-flags is undesirable, but that exactly is my plan "C".

> b) it would remove the need to pass list of configs, just pass `config = { imports = cfg.config; }`

Hm, my immediate objections to this would be "but `evalModules` takes a list" and "it would make `types.opaque` handling a bit more complicated", but I'll think about it some more.

> 2. It also not very cool that `pkgs/top-level/default.nix` includes hacky "check only known options".

> The `unknowns` hack is because module system doesn't support partial checks, right?

Yep.

> Yes, I agree this is sole usage, but it may be harder to sell to others.

I think that this hackyness is ok because it is temporary: it will exist only until all the config options are added to `pkgs/top-level/config.nix`. I'm not very exited at the idea of making a generic mechanism for a use case that is designed to be eliminated ASAP.

> Your thoughts about overlays in config are very interesting!

I think so too :] But the actual implementation needs some more work because ATM the fixpoint over pkgs is made too late in `stage.nix` and `config.nix` is fed with the resulting `pkgs` which means that, e.g.

```
{ pkgs, ... }:
{
  config = {
    packageOverrides = _: {
      hello-test = pkgs.hello;
    };
  };
}
```

fails with infinite recursion error.

I'm still considering if I should fix that in this branch or not.

**Comment by oxij** [comment]
> But the actual implementation needs some more work because ATM the fixpoint over pkgs is made too late in `stage.nix` and `config.nix` is fed with the resulting `pkgs` which means that, e.g.
>
> ```
> { pkgs, ... }:
> {
>   config = {
>     packageOverrides = _: {
>       hello-test = pkgs.hello;
>     };
>   };
> }
> ```
>
> fails with infinite recursion error.

Ha-ha, turns out the problem was with `attrsOf`, not with the fixpointing of `pkgs`. I made a variant of `attrsOf` that is lazy in its attribute values and everything now works.

I.e. `packageOverrides` is now equivalent to an overlay for free. Example:

```
{ pkgs, ... }:
{
  config = {
    packageOverrides = super: {
      libcardiacarrest = super.libcardiacarrest.overrideAttrs (oldAttrs: { name = "ha-ha" + oldAttrs.name; });
      libpulseaudio = pkgs.libcardiacarrest;
      hello-test = pkgs.hello;
    };
  };
}
```

:]

File: pkgs/top-level/default.nix [hunk 1]
```diff
@@ -41,45 +41,59 @@
...
-    builtins.intersectAttrs { platform = null; } config1
+    builtins.intersectAttrs { platform = null; } config
     // args.localSystem);
 
   crossSystem = if crossSystem0 == null then localSystem
                 else lib.systems.elaborate crossSystem0;
 
+  # Massage e into a NixOS module.
+  # We need all this stuff only because we want to support unknown options,
+  # without them this can be simplified a lot.
```
Comment (Infinisil) [inline pkgs/top-level/default.nix]: I think this should also indicate how unknown options are not something to stay, which can justify this hacky implementation.

File: lib/types.nix [hunk 2]
```diff
@@ -291,6 +291,15 @@ rec {
       functor = (defaultFunctor name) // { wrapped = elemType; };
     };
 
+    # Same as attrsOf, but lazy towards attribute values (at the cost of not supporting mkIf and etc)
+    lazyAttrsOf = elemType: attrsOf elemType // {
```
Comment (Infinisil) [inline lib/types.nix]: I'm not sure if it's worth introducing this confusing and slightly module breaking type (because it makes `mkIf`, etc. not work anymore), just to add support for using the slightly deprecated `packageOverrides` as overlays. If people want overlays they can use overlays directly, if the package set they're using doesn't support this then that should be fixed there imo.

**Review by infinisil — CHANGES_REQUESTED** [review]
Very cool overall!

File: pkgs/top-level/config.nix [hunk 3]
```diff
@@ -26,7 +26,12 @@ let
     '';
   });
 
-  options = {
+  mkOverrides = args: mkMassRebuild ({
+    type = types.functionTo (types.lazyAttrsOf (types.uniq types.unspecified));
+    default = super: {};
```
Comment (Infinisil) [inline pkgs/top-level/config.nix]: Actually I don't think we need all these new types by just not supporting setting `packageOverrides` multiple times, which is totally okay because this wasn't supported previously either. `packageOverrides`-like ways are kind of deprecated anyways too, we don't need to extend their support by giving new features. If people want to have overlay-like behavior they should use overlays, which package sets can add support for too (e.g. I'm trying to get python to support overlays right now in https://github.com/NixOS/nixpkgs/pull/67422). In addition, this type introduces yet another slightly different way of overriding packages, different from any other ways. I think we have enough ways of overriding things already, this will only make it more confusing.

So in short, I think this can be `types.uniq types.unspecified` without worry, then the controversial `functionTo` and `lazyAttrsOf` can be dropped from this PR.

**Comment by oxij** [comment]
> Actually I don't think we need all these new types by just not supporting setting `packageOverrides` multiple times, which is totally okay because this wasn't supported previously either.

That's not the only problem it solves, it also allows to do stuff like https://github.com/NixOS/nixpkgs/pull/57123#issuecomment-472950968

> `packageOverrides`-like ways are kind of deprecated anyways too, we don't need to extend their support by giving new features

Weeell, you see, IMHO the example in https://github.com/NixOS/nixpkgs/pull/57123#issuecomment-472950968, as noted above, actually gets the whole `overlays` thing into quite a pickle.

With properly typed `packageOverrides`, as in this PR, you don't really need `overlays`, you can just tie the knot over `pkgs` as https://github.com/NixOS/nixpkgs/pull/57123#issuecomment-472950968 does, which makes `packageOverrides` equivalent to `overlays`.

Moreover, consider this:

- the `overlays` attribute of nixpkgs seems like a part of `config` to me (it configures nixpkgs after all, moreover `grep 'inherit.*overlays'` shows that it _always_ gets inherited together with `config`, thus, IMO, it should be a part of `config`),
- `config` already has `packageOverrides`,
- which, given these changes, can express the same things as `overlays` without `overlays`.

That is to say, the `overlays` attribute of nixpkgs is just a list of `functionsTo (functionsTo (lazyAttrsOf whatever))` that gets merged and tied in a recursive knot using custom code instead of the normal `merge` function of the `lib.types` type it, semantically, has.

Therefore, IMHO, at the very least, we should move `overlays` into `config` (thus, probably, requiring all these new types, unless there's a way to fix the default ones to be more lazy) making `overlays` and `packageOverrides` into slightly different variants of the same thing. Either one of those can then be deprecated, but it is not so obvious to me which one should be deprecated, since by following a bit extended suggestion by @danbst above, we can just make all the files in `~/.nixpkgs/config/` into NixOS modules, thus getting everything else that `overlays` do with custom code for absolutely free for both NixOS and non-NixOS uses of Nixpkgs.

The rest are good suggestions, I'll use them (and the ones by @danbst) when making the v2 of this. Thanks.

**Comment by Infinisil** [comment]
A problem with this is that this doesn't give you the full power of overlays. I think with how it's defined right now, it's impossible to do multiple changes to the same attribute in different `packageOverrides`, which standard overlays support. However it might be possible to weave the different layers of the overlays through the merge function, so this can maybe even gotten to work.

But a big problem with this I'm just seeing is that this will be very unnatural and maybe problematic for other package sets. E.g. suppose you want to override some python package in an overlay-style fashion and you decide to use this new way of defining `pythonPackageOverrides` in `config.nix`. What is `self` going to be then? Is it `python.pkgs`/`pythonPackages` or perhaps `python3Packages` because a different overlay changed the default or so? And maybe some program even wants a program-specific overlay, which doesn't get propagated to the `pkgs` the module receives, then the package set from `pkgs` would be very wrong.

I guess the thing I'm getting at here is that `pkgs` is a like a global variable, it's shared accross everything and it would be the only way to access the result of the `pkgs` fixed-point. Overlays in comparison have a locally scoped `self`, which allows it to be much more flexible. And the `self: super:` structure lends itself well to uses outside of the module system too, see `lib.makeExtensible` and co.

**Comment by Infinisil** [comment]
In fact I think the python example doesn't actually have a correct answer, because the overlay would get used for both python2 and python3

**Comment by oxij** [comment]
> A problem with this is that this doesn't give you the full power of overlays. I think with how it's defined right now, it's impossible to do multiple changes to the same attribute in different `packageOverrides`, which standard overlays support.

Sure, but that's just because I made this into a simple demonstration of `functionTo`. I could have made `packageOverrides` into `types.opaque` and concatMapped that list to `overlays` to get exactly the opposite effect, for instance.

(But to be honest, I'm very suspicious of the idea of overriding the same thing differently in different modules and expecting anything good to come out of it. Which is why I use `types.uniq` there.)

> However it might be possible to weave the different layers of the overlays through the merge function, so this can maybe even gotten to work.

Yes, that is the point. In fact, if `overlays` are moved into that module, the problem gets substantially easier.

> But a big problem with this I'm just seeing is that this will be very unnatural and maybe problematic for other package sets. E.g. suppose you want to override some python package in an overlay-style fashion and you decide to use this new way of defining `pythonPackageOverrides` in `config.nix`. What is `self` going to be then? Is it `python.pkgs`/`pythonPackages` or perhaps `python3Packages` because a different overlay changed the default or so?

Even as it is now, in principle, nothing prevents you from doing this:

```
{ pkgs, ... }:
{
  config = {
    python3PackageOverrides = super: {
      whatever = pkgs.python3Packages.whateverElse.overrideAttrs (...);
    };
  };
}
```

But, yes, ideally `pythonPackageOverrides` and such they should take their own set as `super` (not `pkgs` like `perlPackageOverrides` does, for instance), similarly to `overlays`. Then, even without the versioned ones you could do something like this:

```
{ pkgs, ... }:
{
  config = {
    pythonPackageOverrides = super: 
      let pythonPackages = if !super.isPy3k then pkgs.python2Packages else pkgs.python3Packages;
    in {
      whateverElse = super.whateverElse.overrideAttrs (...);
      whatever = pythonPackages.whateverElse;
    };
  };
}
```

But I think we should just have both `pythonPackageOverrides` and the versioned ones (which is exactly why I wanted #40417).

**Comment by Infinisil** [comment]
The way you showed there, using `isPy3k` to decide where to get `self` from, is really not very good, because it's easy to make mistakes and get a bad override. Having to point out that everybody needs to remember to add some `isPy3k` boilerplate to their overrides otherwise they won't work seems like a step down. And what about package sets with 10 different versions? This doesn't scale well at all.

I'd really like for this "making `packageOverrides` work as overlays" functionality to be moved into a separate PR, it's the only change here that's rather controversial, and it can be separated cleanly from here.

Also pinging @nbp who probably has an opinion on this.

**Comment by nbp** [comment]
A 5 seconds reply, before a larger one coming over the weekend.

> ### _file -> file
> 
> I've used `_file` for my custom modules already. There has to be backwards compat. But I also wonder why was it named like that. Maybe, to discourage usage?

I named it `_file`, because we have to carry the source information. The problem of naming it `file` is that `file` is very likely to be a valid attribute set or option name.

What you do not see is that this attribute is propagated at all the level of the evaluation, and will cause issues as soon as a single option is named file. Either you will accidentally define the option, or you will take the value defined as argument as a file location.

`_file` is meant, as the underscore suggests, to be an artifact of the module system, not a valid option name.

**Comment by Infinisil** [comment]
@nbp is the longer review still coming?

**Comment by Infinisil** [comment]
My https://github.com/NixOS/nixpkgs/pull/75031 includes making `unifyModuleSyntax` idempotent the other way around, with `file -> _file` (where necessary)
